README.md
# Kasai Core
A Node.js library that manages the process lifecycle and orchestrates the
registration, activation, and termination of plugins. It provides logging as
standard output in JSON format and exception reporting to [Sentry][sentry].

[API Documentation](https://version36.gitlab.io/kasai/core)

[Kasai][kasai] is a suite of Node.js libraries for building web services.

[kasai]: https://gitlab.com/version36/kasai "Kasai"
[sentry]: https://sentry.io "Sentry"


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://gitlab.com/version36/kasai/core/compare/v0.1.1...v0.1.2) (2020-02-16)

### [0.1.1](https://gitlab.com/version36/kasai/core/compare/v0.1.0...v0.1.1) (2020-02-16)

## 0.1.0 (2020-02-15)


__tests__/env.spec.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

import test from 'ava'

test.beforeEach(() => {
  delete require.cache[require.resolve('../lib/env')]
})

test.serial('Environment Defined', async spec => {
  process.env.NODE_ENV = 'env'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'env')
  spec.false(env.DEVELOPMENT)
  spec.false(env.TEST)
  spec.false(env.SANDBOX)
  spec.false(env.STAGING)
  spec.false(env.PRODUCTION)
})

test.serial('Environment in Development', async spec => {
  process.env.NODE_ENV = 'development'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'development')
  spec.true(env.DEVELOPMENT)
})

test.serial('Environment in Test', async spec => {
  process.env.NODE_ENV = 'test'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'test')
  spec.true(env.TEST)
})

test.serial('Environment in Sandbox', async spec => {
  process.env.NODE_ENV = 'sandbox'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'sandbox')
  spec.true(env.SANDBOX)
})

test.serial('Environment in Staging', async spec => {
  process.env.NODE_ENV = 'staging'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'staging')
  spec.true(env.STAGING)
})

test.serial('Environment in Production', async spec => {
  process.env.NODE_ENV = 'production'
  const env = await import('../lib/env')
  spec.is(env.ENV, 'production')
  spec.true(env.PRODUCTION)
})

test.serial('Environment Debug Enabled', async spec => {
  const values = ['1', 'active', 'enabled', 'on', 'true', 'y', 'yes']
  for (const value of values) {
    process.env.NODE_DEBUG = value
    const env = await import('../lib/env')
    spec.true(env.DEBUG)
  }
})

test.serial('Environment Debug Disabled', async spec => {
  process.env.NODE_DEBUG = 'other'
  const env = await import('../lib/env')
  spec.false(env.DEBUG)
})


__tests__/error.spec.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

import test from 'ava'
import { ServiceError } from '../lib/error'

test('Create Service Error', spec => {
  let error = new ServiceError('message')
  const { code, stack } = error
  spec.true(error instanceof Error)
  spec.true(typeof stack === 'string')
  spec.is(code, undefined)
  spec.is(error.message, 'message')

  let exception = new Error('exception')
  error = new ServiceError(exception)
  spec.true(error instanceof Error)
  spec.is(error.message, 'exception')

  exception = { message: 'literal' }
  error = new ServiceError(exception)
  spec.true(error instanceof Error)
  spec.is(error.message, 'literal')

  exception = { source: 'test' }
  error = new ServiceError(exception)
  spec.true(error instanceof Error)
  spec.is(error.message, 'An unspecified error has occurred.')

  spec.true(new ServiceError('message', 400) instanceof Error)

  spec.true(new ServiceError('message', 500) instanceof Error)

  exception = { code: 'INVALID_REQUEST' }
  spec.true(new ServiceError(exception, 400) instanceof Error)

  exception = { stack: 'stack' }
  spec.true(new ServiceError(exception) instanceof Error)

  exception = { status: 500 }
  spec.true(new ServiceError(exception) instanceof Error)

  exception = { statusCode: 500 }
  spec.true(new ServiceError(exception) instanceof Error)

  exception = { status_code: 500 }
  spec.true(new ServiceError(exception) instanceof Error)

  spec.true(new ServiceError(false) instanceof Error)

  spec.throws(() => new ServiceError([1, 2]))
})

test('Adding Labels', spec => {
  const error = new ServiceError('message')
    .addLabels('A', 'B')
    .addLabels('c', 'D')
    .addLabels('d', 'b')
  const labels = Array.from(error.labels)
  spec.deepEqual(labels, ['A', 'B', 'C', 'D'])
})

test('Adding Meta', spec => {
  const error = new ServiceError('message')
    .addMeta('alpha', { one: 1 })
    .addMeta('alpha', { two: 2 })
    .addMeta('bravo', 'first')
    .addMeta('bravo', 'second')
  spec.deepEqual(error.meta, { alpha: { one: 1, two: 2 }, bravo: 'second' })
})

test('Bad Request', spec => {
  const error = ServiceError.badRequest('message')
  spec.is(error.status, 400)
})

test('Unauthorized', spec => {
  const error = ServiceError.unauthorized('message')
  spec.is(error.status, 401)
})

test('Payment Required', spec => {
  const error = ServiceError.paymentRequired('message')
  spec.is(error.status, 402)
})

test('Forbidden', spec => {
  const error = ServiceError.forbidden('message')
  spec.is(error.status, 403)
})

test('Not Found', spec => {
  const error = ServiceError.notFound('message')
  spec.is(error.status, 404)
})

test('Method Not Allowed', spec => {
  const error = ServiceError.methodNotAllowed('message', ['GET', 'POST'])
  const { allowed } = error.meta
  spec.is(error.status, 405)
  spec.deepEqual(allowed, ['GET', 'POST'])
})

test('Not Acceptable', spec => {
  const error = ServiceError.notAcceptable('message')
  spec.is(error.status, 406)
})

test('Proxy Auth Required', spec => {
  const error = ServiceError.proxyAuthRequired('message')
  spec.is(error.status, 407)
})

test('Request Timeout', spec => {
  const error = ServiceError.requestTimeout('message')
  spec.is(error.status, 408)
})

test('Conflict', spec => {
  const error = ServiceError.conflict('message')
  spec.is(error.status, 409)
})

test('Gone', spec => {
  const error = ServiceError.gone('message')
  spec.is(error.status, 410)
})

test('Length Required', spec => {
  const error = ServiceError.lengthRequired('message')
  spec.is(error.status, 411)
})

test('Precondition Failed', spec => {
  const error = ServiceError.preconditionFailed('message')
  spec.is(error.status, 412)
})

test('Payload Too Large', spec => {
  const error = ServiceError.payloadTooLarge('message')
  spec.is(error.status, 413)
})

test('URI Too Long', spec => {
  const error = ServiceError.uriTooLong('message')
  spec.is(error.status, 414)
})

test('Unsupported Media Type', spec => {
  const error = ServiceError.unsupportedMediaType('message')
  spec.is(error.status, 415)
})

test('Range Not Satisfiable', spec => {
  const error = ServiceError.rangeNotSatisfiable('message')
  spec.is(error.status, 416)
})

test('Expectation Failed', spec => {
  const error = ServiceError.expectationFailed('message')
  spec.is(error.status, 417)
})

test('Kermit Drinks Tea', spec => {
  const error = ServiceError.kermitDrinksTea('message')
  spec.is(error.status, 418)
})

test('Misdirected Request', spec => {
  const error = ServiceError.misdirectedRequest('message')
  spec.is(error.status, 421)
})

test('Unprocessable Entity', spec => {
  const error = ServiceError.unprocessableEntity('message')
  spec.is(error.status, 422)
})

test('Locked', spec => {
  const error = ServiceError.locked('message')
  spec.is(error.status, 423)
})

test('Failed Dependency', spec => {
  const error = ServiceError.failedDependency('message')
  spec.is(error.status, 424)
})

test('Unordered Collection', spec => {
  const error = ServiceError.unorderedCollection('message')
  spec.is(error.status, 425)
})

test('Upgrade Required', spec => {
  const error = ServiceError.upgradeRequired('message')
  spec.is(error.status, 426)
})

test('Precondition Required', spec => {
  const error = ServiceError.preconditionRequired('message')
  spec.is(error.status, 428)
})

test('Too Many Requests', spec => {
  const error = ServiceError.tooManyRequests('message')
  spec.is(error.status, 429)
})

test('Fields Too Large', spec => {
  const error = ServiceError.fieldsTooLarge('message')
  spec.is(error.status, 431)
})

test('Legally Unavailable', spec => {
  const error = ServiceError.legallyUnavailable('message')
  spec.is(error.status, 451)
})

test('Internal Server Error', spec => {
  const error = ServiceError.internalServerError('message')
  spec.is(error.status, 500)
})

test('Not Implemented', spec => {
  const error = ServiceError.notImplemented('message')
  spec.is(error.status, 501)
})

test('Bad Gateway', spec => {
  const error = ServiceError.badGateway('message')
  spec.is(error.status, 502)
})

test('Service Unavailable', spec => {
  const error = ServiceError.serviceUnavailable('message')
  spec.is(error.status, 503)
})

test('Gateway Timeout', spec => {
  const error = ServiceError.gatewayTimeout('message')
  spec.is(error.status, 504)
})

test('.unsupportedHttpVersion', spec => {
  const error = ServiceError.unsupportedHttpVersion('message')
  spec.is(error.status, 505)
})

test('Variant Also Negotiates', spec => {
  const error = ServiceError.variantAlsoNegotiates('message')
  spec.is(error.status, 506)
})

test('Insufficient Storage', spec => {
  const error = ServiceError.insufficientStorage('message')
  spec.is(error.status, 507)
})

test('Loop Detected', spec => {
  const error = ServiceError.loopDetected('message')
  spec.is(error.status, 508)
})

test('Bandwidth Exceeded', spec => {
  const error = ServiceError.bandwidthExceeded('message')
  spec.is(error.status, 509)
})

test('Not Extended', spec => {
  const error = ServiceError.notExtended('message')
  spec.is(error.status, 510)
})

test('Network Auth Required', spec => {
  const error = ServiceError.networkAuthRequired('message')
  spec.is(error.status, 511)
})


__tests__/utility.spec.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

import test from 'ava'
import * as utility from '../lib/utility'

test('Value is Literal Object', spec => {
  spec.true(utility.isLiteral({}))
  spec.false(utility.isLiteral(undefined))
  spec.false(utility.isLiteral(null))
  spec.false(utility.isLiteral(false))
  spec.false(utility.isLiteral(true))
  spec.false(utility.isLiteral(0))
  spec.false(utility.isLiteral('0'))
  spec.false(utility.isLiteral([0]))
  spec.false(utility.isLiteral(Symbol('test')))
  spec.false(utility.isLiteral(new Error('test')))
})

test('Value is Primitive', spec => {
  spec.true(utility.isPrimitive(undefined))
  spec.true(utility.isPrimitive(null))
  spec.true(utility.isPrimitive(false))
  spec.true(utility.isPrimitive(true))
  spec.true(utility.isPrimitive(0))
  spec.true(utility.isPrimitive('0'))
  spec.true(utility.isPrimitive(Symbol('test')))
  spec.false(utility.isPrimitive({}))
  spec.false(utility.isPrimitive([0]))
  spec.false(utility.isPrimitive(new Error('test')))
})

test('Merge Objects', spec => {
  const a = {
    c: [1, 2, 3],
    d: { m: 1 },
    e: { o: 3 },
    x: 1,
    y: 2
  }
  const b = {
    c: [3, 4, 5],
    d: { n: 2 },
    e: { o: 3 },
    y: 3,
    z: 4
  }
  spec.deepEqual(
    utility.merge(a, b),
    {
      c: [1, 2, 3, 4, 5],
      d: { m: 1, n: 2 },
      e: { o: 3 },
      x: 1,
      y: 3,
      z: 4
    }
  )
})

test('Merge Arrays', spec => {
  const a = [1, 2, 3, 4, 5]
  const b = [3, 4, 5, 6, 7]
  const c = [5, 6, 7, 8, 9]
  spec.deepEqual(
    utility.mergeLists(a, b, c),
    [1, 2, 3, 4, 5, 6, 7, 8, 9]
  )
})


lib/env.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

const {
  NODE_ENV,
  NODE_DEBUG
} = process.env

/**
 * Parses environment variable as boolean.
 * @param {?string} env - Environment Variable
 * @returns {boolean} Result
 * @inner
 */
function parseBoolean (env) {
  const values = ['1', 'active', 'enabled', 'on', 'true', 'y', 'yes']
  return values.includes(String(env).toLowerCase())
}

/**
 * Represents current environment.
 * @type {string}
 */
export const ENV = NODE_ENV

/**
 * Represents if environment is in debug mode.
 * @type {boolean}
 */
export const DEBUG = parseBoolean(NODE_DEBUG)

/**
 * Represents if environment is in development.
 * @type {boolean}
 */
export const DEVELOPMENT = NODE_ENV === 'development'

/**
 * Represents if environment is in testing.
 * @type {boolean}
 */
export const TEST = NODE_ENV === 'test'

/**
 * Represents if environment is in a sandbox.
 * @type {boolean}
 */
export const SANDBOX = NODE_ENV === 'sandbox'

/**
 * Represents if environment is in staging.
 * @type {boolean}
 */
export const STAGING = NODE_ENV === 'staging'

/**
 * Represents if environment is in production.
 * @type {boolean}
 */
export const PRODUCTION = NODE_ENV === 'production'


lib/utility.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

/**
 * Returns whether value is a literal object or not.
 * @param {*} any - Any Value
 * @returns {boolean} Result
 */
export function isLiteral (any) {
  return Object.prototype.toString.call(any) === '[object Object]'
}

/**
 * Returns whether value is a primitive or not.
 * @param {*} any - Any Value
 * @returns {boolean} Result
 */
export function isPrimitive (any) {
  return any === null ||
    (typeof any !== 'object' && typeof any !== 'function')
}

/**
 * Merges two object literals.
 * @param {!Object} a - First Object
 * @param {!Object} b - Second Object
 * @returns {Object} Merged Object
 */
export function merge (a, b) {
  const n = {}
  mergeLists(Object.keys(a), Object.keys(b)).forEach(k => {
    if (k in a && !(k in b)) {
      n[k] = a[k]
      return
    }
    if (k in b && !(k in a)) {
      n[k] = b[k]
      return
    }
    const vA = a[k]
    const vB = b[k]
    if (vA === vB) {
      n[k] = vA
      return
    }
    if (Array.isArray(vA) && Array.isArray(vB)) {
      n[k] = mergeLists(vA, vB)
      return
    }
    if (isLiteral(vA) && isLiteral(vB)) {
      n[k] = merge(vA, vB)
      return
    }
    n[k] = vB
  })
  return n
}

/**
 * Merges arrays or takes single array and returns one
 * without duplicate primitives or references.
 * @param {...!Array} lists - Lists
 * @returns {Array} Merged List
 */
export function mergeLists (...lists) {
  return [...new Set([].concat(...lists))]
}


lib/_bootstrap.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

import ms from 'ms'
import { DEVELOPMENT } from './env'
import { LEVELS, STATES, system } from './system'

/**
 * Gracefully shutdowns service on signal event.
 * @param {'SIGHUP'|'SIGINT'|'SIGTERM'} signal - Signal
 * @returns {Promise<void>}
 * @listens NodeJS.Process~SIGHUP
 * @listens NodeJS.Process~SIGINT
 * @listens NodeJS.Process~SIGTERM
 * @inner
 */
async function shutdown (signal) {
  let message
  switch (signal) {
    case 'SIGHUP': {
      message = `${system} is being interrupted.`
      break
    }
    case 'SIGINT': {
      message = `${system} is being interrupted.`
      break
    }
    case 'SIGTERM': {
      message = `${system} is being interrupted.`
      break
    }
  }
  system.report(message, LEVELS.WARNING)
  if (system.state === STATES.RUNNING) await system.stop()
  await system.close()
}

process.once('SIGINT', shutdown)
process.once('SIGTERM', shutdown)
process.once('SIGHUP', shutdown)

/* istanbul ignore next */
if (DEVELOPMENT) {
  // Emitted by Nodemon to restart process.
  process.once('SIGUSR2', async function handleRestart () {
    system.warn(`${system} is being restarted.`, 'SIGUSR2')
    if (system.state === STATES.RUNNING) await system.stop()
    await system.close()
    process.kill(process.pid, 'SIGUSR2')
  })
}

/* istanbul ignore next */
process.once('beforeExit',
  /**
   * @external NodeJS
   * @ignore
   */
  /**
   * @external Process
   * @memberof NodeJS
   * @ignore
   */
  /**
   * @event beforeExit
   * @memberof NodeJS.Process
   * @ignore
   */
  /**
   * Handles `beforeExit` event.
   * @returns {Promise<void>}
   * @listens NodeJS.Process~event:beforeExit
   * @inner
   */
  async function handleBeforeExit () {
    await system.close()
  })

process.on('warning',
  /**
   * Handles `warning` event.
   * @param {Error} warning - Warning Exception
   * @listens NodeJS.Process~warning
   * @inner
   */
  function handleWarning (warning) {
    system.report(warning, LEVELS.WARNING)
  })

process.on('exit',
  /**
   * Handles `exit` event.
   * @param {number} code - Exit Code
   * @listens NodeJS.Process~exit
   * @inner
   */
  function handleExit (code) {
    const uptime = process.uptime()
    const period = ms(process.uptime() * 1000, { long: true })
    const exited = code ? `exited with code ${code}` : 'exited'
    const message = `${system} has ${exited} after ${period} of uptime.`
    if (code) system.warn({ uptime, code }, message, 'exit')
    else system.info({ uptime }, message, 'exit')
  })


lib/error.js
/*
 * Copyright 2020 Robert C Degraffenreaidt
 * Licensed under the MIT License.
 */

import { isLiteral, merge } from './utility'

/**
 * Standardized error for use with reporting and HTTP responses.
 * @extends Error
 * @param {!*} exception - Exception
 * @param {!number} [status] - HTTP Status Code
 */
export class ServiceError extends Error {
  constructor (exception, status) {
    if (Array.isArray(exception)) {
      throw new TypeError(
        '`ServiceError` does not support an array as an exception.'
      )
    }

    super()

    /**
     * Code
     * @type {string}
     * @instance
     */
    this.code = undefined

    /**
     * Labels
     * @type {Set<string>}
     * @instance
     */
    this.labels = new Set()

    /**
     * Message
     * @type {string}
     * @instance
     */
    this.message = 'An unspecified error has occurred.'

    /**
     * Metadata
     * @type {Object}
     * @instance
     */
    this.meta = {}

    /**
     * Name
     * @type {string}
     * @instance
     */
    this.name = 'ServiceError'

    // Used by AVA (test runner). Closure Compiler inspection throws
    // a warning when property is not implemented.
    this.parent = Error

    /**
     * Stack Trace
     * @type {string}
     * @instance
     */
    this.stack = undefined

    /**
     * HTTP Status Code
     * @type {number}
     * @instance
     */
    this.status = status

    if (exception instanceof Error) {
      Object.defineProperties(this, Object.getOwnPropertyDescriptors(exception))
    } else {
      Error.call(this)
      Error.captureStackTrace(this, this.constructor)
    }

    if (typeof exception === 'object') {
      const { code, message, stack, ...data } = exception
      if (message && typeof message === 'string') this.message = message
      if (stack && typeof stack === 'string') this.stack = stack
      if (code && typeof code === 'string') this.code = code

      const {
        status: _status, statusCode, status_code: _statusCode, ...meta
      } = data
      if (typeof this.status === 'undefined') {
        this.status = typeof _status === 'number' ? _status
          : typeof statusCode === 'number' ? statusCode
            : typeof _statusCode === 'number' ? _statusCode
              : 500
      }
      this.meta = meta
    } else {
      if (exception && typeof exception === 'string') this.message = exception
      if (typeof this.status === 'undefined') this.status = 500
    }
  }

  /**
   * Creates Bad Request error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 400 Error
   */
  static badRequest (exception) {
    return new this(exception, 400)
  }

  /**
   * Creates Unauthorized error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 401 Error
   */
  static unauthorized (exception) {
    return new this(exception, 401)
  }

  /**
   * Creates Payment Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 402 Error
   */
  static paymentRequired (exception) {
    return new this(exception, 402)
  }

  /**
   * Creates Forbidden error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 403 Error
   */
  static forbidden (exception) {
    return new this(exception, 403)
  }

  /**
   * Creates Not Found error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 404 Error
   */
  static notFound (exception) {
    return new this(exception, 404)
  }

  /*
   * Method Not Allowed
   * @param {!*} exception - Exception
   * @param {!string[]} allowed - Allowed HTTP Methods
   * @returns {ServiceError} 405 Error
   */
  /**
   * Method Not Allowed
   * @param {!*} exception - Exception
   * @param {string[]} allowed - Allowed HTTP Methods
   * @returns {ServiceError} 405 Error
   */
  static methodNotAllowed (exception, allowed) {
    return new this(exception, 405).addMeta('allowed', allowed)
  }

  /**
   * Creates Not Acceptable error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 406 Error
   */
  static notAcceptable (exception) {
    return new this(exception, 406)
  }

  /**
   * Creates Proxy Authentication Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 407 Error
   */
  static proxyAuthRequired (exception) {
    return new this(exception, 407)
  }

  /**
   * Creates Request Timeout error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 408 Error
   */
  static requestTimeout (exception) {
    return new this(exception, 408)
  }

  /**
   * Creates Conflict error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 409 Error
   */
  static conflict (exception) {
    return new this(exception, 409)
  }

  /**
   * Creates Gone error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 410 Error
   */
  static gone (exception) {
    return new this(exception, 410)
  }

  /**
   * Creates Length Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 411 Error
   */
  static lengthRequired (exception) {
    return new this(exception, 411)
  }

  /**
   * Creates Precondition Failed error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 412 Error
   */
  static preconditionFailed (exception) {
    return new this(exception, 412)
  }

  /**
   * Creates Payload Too Large error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 413 Error
   */
  static payloadTooLarge (exception) {
    return new this(exception, 413)
  }

  /**
   * Creates URI Too Long error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 414 Error
   */
  static uriTooLong (exception) {
    return new this(exception, 414)
  }

  /**
   * Creates Unsupported Media Type error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 415 Error
   */
  static unsupportedMediaType (exception) {
    return new this(exception, 415)
  }

  /**
   * Creates Range Not Satisfiable error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 416 Error
   */
  static rangeNotSatisfiable (exception) {
    return new this(exception, 416)
  }

  /**
   * Creates Expectation Failed error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 417 Error
   */
  static expectationFailed (exception) {
    return new this(exception, 417)
  }

  /**
   * Creates Kermit Drinks Tea ("I'm a Teapot") error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 418 Error
   */
  static kermitDrinksTea (exception) {
    return new this(exception, 418)
  }

  /**
   * Creates Misdirected Request error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 421 Error
   */
  static misdirectedRequest (exception) {
    return new this(exception, 421)
  }

  /**
   * Creates Unprocessable Entity error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 422 Error
   */
  static unprocessableEntity (exception) {
    return new this(exception, 422)
  }

  /**
   * Creates Locked error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 423 Error
   */
  static locked (exception) {
    return new this(exception, 423)
  }

  /**
   * Creates Failed Dependency error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 424 Error
   */
  static failedDependency (exception) {
    return new this(exception, 424)
  }

  /**
   * Creates Unordered Collection error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 425 Error
   */
  static unorderedCollection (exception) {
    return new this(exception, 425)
  }

  /**
   * Creates Upgrade Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 426 Error
   */
  static upgradeRequired (exception) {
    return new this(exception, 426)
  }

  /**
   * Creates Precondition Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 428 Error
   */
  static preconditionRequired (exception) {
    return new this(exception, 428)
  }

  /**
   * Creates Too Many Requests error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 429 Error
   */
  static tooManyRequests (exception) {
    return new this(exception, 429)
  }

  /**
   * Creates Request Header Fields Too Large error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 431 Error
   */
  static fieldsTooLarge (exception) {
    return new this(exception, 431)
  }

  /**
   * Creates Unavailable For Legal Reasons error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 451 Error
   */
  static legallyUnavailable (exception) {
    return new this(exception, 451)
  }

  /**
   * Creates Internal Server Error error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 500 Error
   */
  static internalServerError (exception) {
    return new this(exception, 500)
  }

  /**
   * Creates Not Implemented error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 501 Error
   */
  static notImplemented (exception) {
    return new this(exception, 501)
  }

  /**
   * Creates Bad Gateway error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 502 Error
   */
  static badGateway (exception) {
    return new this(exception, 502)
  }

  /**
   * Creates Service Unavailable error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 503 Error
   */
  static serviceUnavailable (exception) {
    return new this(exception, 503)
  }

  /**
   * Creates Gateway Timeout error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 504 Error
   */
  static gatewayTimeout (exception) {
    return new this(exception, 504)
  }

  /**
   * Creates HTTP Version Not Supported error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 505 Error
   */
  static unsupportedHttpVersion (exception) {
    return new this(exception, 505)
  }

  /**
   * Creates Variant Also Negotiates error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 506 Error
   */
  static variantAlsoNegotiates (exception) {
    return new this(exception, 506)
  }

  /**
   * Creates Insufficient Storage error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 507 Error
   */
  static insufficientStorage (exception) {
    return new this(exception, 507)
  }

  /**
   * Creates Loop Detected error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 508 Error
   */
  static loopDetected (exception) {
    return new this(exception, 508)
  }

  /**
   * Creates Bandwidth Limit Exceeded error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 509 Error
   */
  static bandwidthExceeded (exception) {
    return new this(exception, 509)
  }

  /**
   * Creates Not Extended error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 510 Error
   */
  static notExtended (exception) {
    return new this(exception, 510)
  }

  /**
   * Creates Network Authentication Required error.
   * @param {!*} exception - Exception
   * @returns {ServiceError} 511 Error
   */
  static networkAuthRequired (exception) {
    return new this(exception, 511)
  }

  /**
   * Adds labels.
   * @param {...!string} keywords - Keywords
   * @returns {ServiceError} Self
   */
  addLabels (...keywords) {
    keywords = keywords.map(keyword => {
      return keyword.trim().toUpperCase().replace(/\W+/g, '_')
    })
    keywords.forEach(keyword => this.labels.add(keyword))
    return this
  }

  /**
   * Adds metadata.
   * @param {!string} name - Name
   * @param {!*} data - Data
   * @returns {ServiceError} Self
   */
  addMeta (name, data) {
    if (name in this.meta && isLiteral(this.meta[name]) && isLiteral(data)) {
      this.meta[name] = merge(this.meta[name], data)
    } else {
      this.meta[name] = data
    }
    return this
  }
}
